applications/LinearSolversApplication/custom_solvers/mkl_ilu.cpp

```diff
@@ -0,0 +1,403 @@
+        case SPARSE_STATUS_INVALID_VALUE:
+            return "SPARSE_STATUS_INVALID_VALUE: The input parameters contain an invalid value.";
+        case SPARSE_STATUS_EXECUTION_FAILED:
+            return "SPARSE_STATUS_EXECUTION_FAILED: Execution failed.";
+        case SPARSE_STATUS_INTERNAL_ERROR:
+            return "SPARSE_STATUS_INTERNAL_ERROR: An error in algorithm implementation occurred.";
+        case SPARSE_STATUS_NOT_SUPPORTED:
+            return "SPARSE_STATUS_NOT_SUPPORTED: The requested operation is not supported.";
+        default:
+            return "unknown";
+    } // switch Status
+}
+
+
+template <class TSparse, class TDense>
+struct MKLILUSmootherBase<TSparse,TDense>::Impl
+{
+    std::vector<typename TSparse::DataType> bilut;
+
+    std::vector<MKL_INT> ibilut;
+
+    std::vector<MKL_INT> jbilut;
+
+    std::array<MKL_INT,128> ipar;
+
+    std::array<typename TSparse::DataType,128> dpar;
+
+    std::vector<typename TSparse::DataType> mBuffer;
+
+    std::unique_ptr<sparse_matrix_t,MKLCSRAdaptorDestructor> mpLUAdaptor;
+
+    MKL_INT mIterations;
+
+    typename TSparse::DataType mRelaxation;
+}; // struct MKLILUSmootherBase::Impl
+
+
+template <class TSparse, class TDense>
+MKLILUSmootherBase<TSparse,TDense>::MKLILUSmootherBase()
+    : MKLILUSmootherBase(Parameters())
+{
+}
+
+
+template <class TSparse, class TDense>
+MKLILUSmootherBase<TSparse,TDense>::MKLILUSmootherBase(Parameters Settings)
+    : Base(),
+      mpImpl(new Impl)
+{
+    KRATOS_TRY
+    Settings.AddMissingParameters(this->GetDefaultParameters());
+    mpImpl->mIterations = Settings["iterations"].Get<int>();
+    mpImpl->mRelaxation = Settings["relaxation"].Get<double>();
+    KRATOS_CATCH("")
+
+    std::fill(mpImpl->ipar.begin(), mpImpl->ipar.end(), static_cast<MKL_INT>(0));
+    std::fill(mpImpl->dpar.begin(), mpImpl->dpar.end(), static_cast<typename TSparse::DataType>(0));
+
+    mpImpl->mpLUAdaptor = nullptr;
+}
+
+
+template <class TSparse, class TDense>
+MKLILUSmootherBase<TSparse,TDense>::~MKLILUSmootherBase() = default;
+
+
+/// Compute the incomplete LU factorization.
+template <class TSparse, class TDense>
+void MKLILUSmootherBase<TSparse,TDense>::InitializeSolutionStep(typename Base::SparseMatrix& rLhs,
+                                                                typename Base::Vector& rSolution,
+                                                                typename Base::Vector& rRhs)
+{
+    KRATOS_TRY
+    Base::InitializeSolutionStep(rLhs, rSolution, rRhs);
+    KRATOS_CATCH("")
+
+    KRATOS_TRY
+    if (!rLhs.nnz()) return;
+
+    // Set factorization parameters.
+    auto& ipar = mpImpl->ipar;
+    auto& dpar = mpImpl->dpar;
+
+    ipar[  4]   = 1e2;      //< maximum number of iterations
+    ipar[  5]   = 1;        //< output error messages
+    ipar[  6]   = 1;        //< output warning messages
+    dpar[ 30]   = 1.0;      //< value to replace small diagonal entries with
```

Comment:
Curious, why are they hard coded?

Reply:
I'd have to provide users access to setting these variables through `Parameters`, which would mean this smoother having a different JSON interface the `ILU` in Core.